text/0000-closure-lifetime-binder.md

```diff
@@ -0,0 +1,266 @@
+- Feature Name: `closure_lifetime_binder`
+- Start Date: 2022-01-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+This RFC went through a pre-RFC phase at https://internals.rust-lang.org/t/pre-rfc-allow-for-a-syntax-with-closures-for-explicit-higher-ranked-lifetimes/15888
+
+# Summary
+
+Allow explicitly specifying lifetimes on closures via `for<'a> |arg: &'a u8| { ... }`. This will always result in a higher-ranked closure which can accept *any* lifetime (as in `fn bar<'a>(val: &'a u8) {}`). Closures defined without the `for<'a>` syntax retain their current behavior: lifetimes will be inferred as either some local region (via an inference variable), or a higher-ranked lifetime.
```

Comment:
Isn't such a closure just rank-1? I think the RFC is great, but for docs "higher-rank" might be misleading.

If I understand correctly,  `for<'a> |arg: &'a u8| { ... }` is the closure version of `fn<'a>(arg: &'a u8) {}`, which is rank 1.
 
 When such a closure is passed to a function `fn takes_closure(f: impl Fn(&u8))` then `takes_closure` (iuc) has a higher-ranked lifetime, but the closure itself does not.

Reply:
I've been using 'higher-ranked' to mean 'has a `for<'a>` binder', in the same way that `for<'a> Type: Trait {}` is called a 'higher-ranked trait.

As far as I know, 'higher-ranked' in the context of Rust is taken to mean 'at least rank 1', instead of 'at least rank 2' (though I could be wrong about this).

Reply:
@Aaron1011 what is a higher-ranked trait? (sorry, I'm relatively new to Rust).

For higher-ranked trait *bounds*, I see two complete examples in the refernece, and they look higher-rank to me:

https://doc.rust-lang.org/stable/reference/trait-bounds.html?highlight=higher-rank#higher-ranked-trait-bounds


```rs
fn call_on_ref_zero<F>(f: F) where for<'a> F: Fn(&'a i32) {
    let zero = 0;
    f(&zero);
}
```

and

```rs
fn call_on_ref_zero<F>(f: F) where F: for<'a> Fn(&'a i32) {
    let zero = 0;
    f(&zero);
}
```

I'm using "higher-rank" in a way that I think is consistent with the higher-ranked trait bounds docs: the type variable is bound to the right of at least one arrow. This is the sense in [Wikipedia](https://en.wikipedia.org/wiki/Parametric_polymorphism#Rank-1_(prenex)_polymorphism) that, according to Types and Programming Languages, goes back to [Leivant, 1983, I think in section 6.2](https://dl.acm.org/doi/pdf/10.1145/567067.567077?casa_token=Vor1-2WkKwgAAAAA:ahrOwmvDs9GyKZUejxSDs30OqRUsPZZiwWLS-1VAmw1bDtOSSps7KPLvAPxEai8gqOrgKsHmhOmiyEc)

Reply:
Closures whose parameters have `forall` are called "higher-rank" because, if they were converted to a `fn` type, that `fn` type would involve a binder.

Reply:
Is `fn id<T>(t: T) -> T { t }` has a binder. Would you consider it higher-ranked? (asking to try to understand better)

Reply:
Pinging on this. Sorry to be pedantic, but I'm genuinely worried about Rust getting stuck with nonstandard terminology, which could be misleading and hard to un-teach.

Much like I often see https://varkor.github.io/blog/2018/07/03/existential-types-in-rust.html referenced, even though I'm pretty sure it's wrong about `impl` always meaning "existential".

Reply:
@mheiber I think `fn id<T>(t: T) -> T { t }` in your example is not considered even rank 1 in Rust, because it's not a first class value. The functions in Rust are monomorphised at compile-time, so the polymorphism is not preserved. (If it was, it would be rank 1) So each "type application" of your example would be rank 0, and the original polymorphic version wouldn't just exist.

However, `for <'a>` is a special case, because lifetimes are equivalent in runtime anyway, so they don't need monomorphisation. That means that the function as a _value_ is actually rank 1 polymorphic.

Reply:
I acknowledge that it may be legitimate to go and try to add a foot-note to disambiguate this, but I also reckon that, imho, it is not that useful to fully forgo that terminology, since `for<'lifetime…>` quantification only really appears when, at some point, a higher-ranked API is involved, which thus makes `for<'lifetime…>` quantifications be either:

  - directly involved in a higher-order signature, as in your `call_on_ref_zero` example;
  - express that a type itself is _compatible with_ a higher-order API.

So maybe a change along the following lines would strike the right balance:

```suggestion
Allow explicitly specifying lifetimes on closures via `for<'a> |arg: &'a u8| { ... }`. This will always result in a higher-ranked[^higher_ranked] closure which can accept *any* lifetime (as in `fn bar<'a>(val: &'a u8) {}`). Closures defined without the `for<'a>` syntax retain their current behavior: lifetimes will be inferred as either some local region (via an inference variable), or a higher-ranked lifetime[^higher_ranked].

[^higher_ranked]: technically this is a misnomer: the closures themselves are not higher-ranked, but rather, rank-1, as in _simply generic over a lifetime_. But that makes them **compatible with higher-ranked APIs**, that is, APIs that expect, themselves, a generic-over-a-lifetime callback. So these `for<'lifetime…>` callback signatures can be labelled as _higher-ranked-compatible_. Moreover, since `for<'lifetime>` quantification is only seen in such cases, then such quantification, in and of itself, ends up dubbed "higher-order" as a shorthand, and by extension, so are the closures featuring it, as well as the lifetime parameters so introduced.
```

Reply:
@golddranks re 

> The functions in Rust are monomorphosed at compile-time so the polymorphism is not preserved ...So each "type application" of your example would be rank 0...

I don't think monomorphization is relevant here. Monomorphization is the process of taking a polymorphic thing and transforming it into non-polymorphic things. Which presumes the thing was polymorphic to begin with.

Reply:
@danielhenrymantilla thanks for suggesting this change, but, my two cents is that it doesn't seem like an improvement to add additional concepts "higher-ranked API" and "compatibility with higher-ranked APIs." 

You wrote:

>  "**compatible with higher-ranked APIs**,  that is, APIs that expect, themselves, a generic-over-a-lifetime callback." 

Is `takes_any` a "higher-ranked" API?

```rs
fn main() {
    let f = for <'_>|| 3;
    takes_any(f);
}

fn takes_any(_t: impl std:any::Any) {}
```

For closures that are not higher-ranked, "higher-ranked" doesn't seem like helpful terminology imo. 

I'm not attached to it, but an example of an alternative is  "closures with lifetime parameters".

Reply:
@mheiber Oops, I started to already replying to you, because GitHub send me an e-mail where you mentioned me, but I guess you deleted it? Let me just share this, because I already made the effort to check whether Haskell is different from Rust here, and turns out it is:

Rust doesn't consider polymorphic types as first-class. They are "proper" types that can have values only after applying the type. This Rust playground demonstrates this; `a`, which is supposed to be a *variable* of type of function `id`,  gets initiated with a `&str`, and doesn't accept an integer after that. (The same happens when you replace `a` with a closure: `let a = |t| t;`)
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0791da7525a893979bc8d97ed78e4f8b

Instead, in Haskell, one can do this: (pardon my Haskell, def. not an expert) `let id = \t -> t in let a = id in (a "hello", a 4)` which works fine. (Apparently https://www.tryhaskell.org doesn't support sharing code examples, but copy-pasting from here works.)

You were absolutely correct that monomorphisation is just an implementation strategy, but clearly Rust has let it also an effect to the semantics of language. To be sure, I think it would be feasible to support "proper" rank-1 types for other than lifetime parameters; this seems feasible for types that don't carry the data of the polymorphic parameter with them, so you can monomorphise at usage site, depending on the usage. The prominent example is having generic closures: e.g. `for<D: Debug> |d: D| println!("{:?}", d);` I think this isn't considered a priority at the moment, but more like "a nice to have some day".

Also, I think that "closures with lifetime parameters" is a more helpful term indeed than "higher ranked", which is jargony, unclear and possibly plain out wrong here.

Reply:
I disagree with "closures with lifetime parameters" terminology, at least, when phrased like that, since it is _waaaay too loose_ of a terminology to help anybody —the horribly unreadable-for-beginners error "one type is more general than the other" stems from this kind of loose "there is a generic lifetime param somewhere" terminology, that then needs to say "whoops, your generic lifetime param is not general enough". Maybe we could add the "`for`-quantified" nuance to these "lifetime parameters", I don't know, what I do know is that something needs to mention that in:

```rust
struct Foo<'a>(*mut Self);

impl<'a> Foo<'a> {
    fn call (_: &'a ()) {}
}

struct Bar(*mut Self);

impl Bar {
    fn call<'a> (_: &'a ()) {}
}
```

both `Foo::call` and `Bar::call` are "(stateless-)closures with a lifetime parameter, `'a`", and yet only `Bar::call` is `for`-quantified / _generically callable_  (whereas `Foo::call` is a _generic [`Foo`'s] callable_).

In other words, there _has_ to be a specific terminology for late-bound lifetimes (there we go, another term[^5]), to distinguish from generics from an outer scope. Moreover, HRTB/ Higher-Rank Trait Bounds is a phrasing that is already part of the language (from [`1.3.0`](https://doc.rust-lang.org/1.3.0/nomicon/hrtb.html) to [now](https://doc.rust-lang.org/1.59.0/nomicon/hrtb.html)).

So "closure that meets/implements a Higher-Rank Trait bound" is official Rust terminology, I think we are way past changing that. Granted, the fact that we then say "higher-rank closure" as a shorthand may be confusing for the more type-theory rigorous people (at least based on this very thread), hence the suggestion of a footnote to soothe that aspect.

[^5]: although there is no user-facing official documentation about early-bound *vs.* late-bound generic lifetimes, so as of now, this would be more confusing than anything else. But that aspect could be independently improved by the official docs, and then the RFC could replace the "higher-rank" terminology with "with late-bound generic lifetimes"? 🤷